File: src/main/java/io/vertigo/shiny/components/form/ShinyFormField.java
package io.vertigo.shiny.components.form;

import io.vertigo.core.lang.Assertion;

import java.util.List;

public record ShinyFormField(
        String name,
        String label,
        ShinyFormFieldType type,
        Object value,
        boolean required,
        String placeholder,
        String helpText,
        Object defaultValue,
        List<ShinyFormOption> options,
        ShinyFormFieldValidator validator,
        boolean readOnly) {

    public ShinyFormField {
        Assertion.check()
                .isNotBlank(name)
                .isNotBlank(label)
                .isNotNull(type);
        // value can be null
        // options can be null if not a SELECT, RADIO, or CHECKBOX_GROUP
    }

    public boolean validate() {
        return validator.validate(value);
    }
}

File: banshee/src/models/ShinyFormField.ts
import { ShinyFormFieldType } from './ShinyFormFieldType';
import { ShinyFormOption } from './ShinyFormOption';
import { ShinyFormFieldValidator } from './ShinyFormFieldValidator';

export class ShinyFormField {
  name: string;
  label: string;
  type: ShinyFormFieldType;
  value?: any;
  required: boolean;
  placeholder?: string;
  helpText?: string;
  defaultValue?: any;
  options?: ShinyFormOption[];
  validator?: ShinyFormFieldValidator;
  readOnly: boolean;

  constructor(
    name: string,
    label: string,
    type: ShinyFormFieldType,
    value: any,
    required: boolean,
    placeholder: string,
    helpText: string,
    defaultValue: any,
    options: ShinyFormOption[],
    validator: ShinyFormFieldValidator,
    readOnly: boolean
  ) {
    this.name = name;
    this.label = label;
    this.type = type;
    this.value = value;
    this.required = required;
    this.placeholder = placeholder;
    this.helpText = helpText;
    this.defaultValue = defaultValue;
    this.options = options;
    this.validator = validator;
    this.readOnly = readOnly;
  }
}
